### DISCUSSION TITLE
Realtime Database stops sending new data after some time.

### DISCUSSION THREAD

**Discussion by dgsdestinygamestudios** [GodotNuts/GodotFirebase#368] [Q&A]
![image](https://github.com/GodotNuts/GodotFirebase/assets/148333601/77dc06f9-65de-447c-a4d5-fe94a0bf2cac)
The number that calculator shows, is how many seconds passed since game started. They go the same (sometimes 1 second off) for some time. Then the player who is waiting for his turn, stops receiving data from Realtime Database.

It happens on seemingly random times. I saw 8 seconds, I saw 113 seconds.

This is the code behind:

```
# The player who currently plays (on the left), calls this code every second.
var database: FirebaseDatabaseReference = Firebase.Database.get_database_reference("rooms")
serialize()
database.update(Game.room.host_username, Game.room.serialize())
```

```
# The player who currently waits (on the right), calls this code every second to see what the other player have done.
_room_data.clear()
var database: FirebaseDatabaseReference = Firebase.Database.get_database_reference("rooms")
database.new_data_update.connect(func(value) -> void: _room_data[value["key"]] = value["data"])
await database.new_data_update # This signal stops firing at some point.
Game.room.deserialize(_room_data[Game.room.host_username])
deserialize(Game.room.board)
```

I would appreciate any help.

**Comment by WolfgangSenff** [comment]
You're using the database all wrong. It's not intended to be pinged like that. Generally, for turn-based games in the realtime database, you want to structure them like this:

0. Both players connect to a MoveList reference in the database, connecting to the new_data_update. Save this reference in the class script, not just at function level.
1. Player 1 makes a move which sends a command to a list with a call to push data. Note this doesn't _change_ player 1's state. A move is just a push to the database.
2. Both players "hear" the update, check the id it was sent with (you'd send the id of the player, plus the command and any arguments needed), and update their state accordingly. This also tells the second player to be allowed to make a move.
3. Player 2 makes a move and repeat.

You can use update as well, but you should _not_ make new references like that. Store one reference and you should be fine.

**Reply by dgsdestinygamestudios** [reply]
I'm using one reference per player. Let me explain what happens now.

Player 1 creates a room. That sends needed data to Realtime Database. I can see the data from the dashboard.
Player 2 enters the room. That updates the data Player 1 sent earlier. I can see the change from the dashboard.

But Player 1's new_data_update never fires. Only Player 2 receives updated data at first, then nothing.

**Reply by WolfgangSenff** [reply]
No you aren't. You're creating a new reference every single time, according to the code you posted. In fact, you're calling both sets of code every second, which means you're making *two* references every second. That's going to eventually blow up in your face. You need to store one reference, and not connect to it with the new signal every single time. Also, if you connect to patch_data_update, when you send an update call, that'll get it every time and give you just the updated info when you do an update. That might not be what you want, though. I do recommend following exactly what I recommended.

**Reply by dgsdestinygamestudios** [reply]
I've declared the reference at the start of the script, then hooked the signal in the _ready() function. Then what I've said happened. I'll try patch_data_update to see if it helps. Thank you for bearing with me. I'll get back to you as soon as I try this advise.

**Reply by WolfgangSenff** [reply]
Regardless of that, you shouldn't be calling anything every second to check stuff - you should be simply letting it do its thing. That's what signals are for! When you push, everyone that is connected should be notified when it gets through to the database and out. I've definitely build multiplayer games that work for longer than 13 seconds, and longer than a minute or whatever. :)

**Reply by WolfgangSenff** [reply]
The whole point of the library is so that, when two (or any amount) players are connected to the same spot in the database, when any one of them pushes data to that database, it pushes out the data to all of them. The signal is what updates you that there's new data. Having each player push in this way directly to that list isn't actually the _safest_ route that can be taken, but for now, it's the right one for you I think, to just try to get it working.

In other words, you *never* have to check after a second once you're connected once. It will always push data out to the players when it receives it, and you will always be updated via your one and only signal connection.
